src/Spice86.Core/Emulator/Devices/ExternalInput/Intel8259Pic.cs

```diff
@@ -0,0 +1,459 @@
+            return true;
+        }
+
+        return GetPriorityPosition(candidate) < GetPriorityPosition(current);
+    }
+
+    /// <summary>
+    ///     Resolves the highest-priority IRQ bit from a pending mask, respecting the rotation baseline.
+    /// </summary>
+    /// <param name="pendingMask">Bitmask of eligible IRQs.</param>
+    /// <returns>The IRQ index or 8 when no candidate is found.</returns>
+    private byte GetHighestPriorityIrq(byte pendingMask) {
+        if (pendingMask == 0) {
+            return 8;
+        }
+
+        byte irq = HighestPriorityIrq;
+        for (int step = 0; step < 8; step++) {
+            byte mask = (byte)(1 << irq);
+            if ((pendingMask & mask) != 0) {
+                return irq;
+            }
+
+            irq = (byte)((irq + 1) & 0x07);
+        }
+
+        return 8;
+    }
+
+    /// <summary>
+    ///     Finds the next IRQ that should be serviced, or 8 when none qualify.
+    /// </summary>
+    internal byte GetNextPendingIrq() {
+        byte pending = (byte)(InterruptRequestRegister & InterruptMaskRegisterInverted & InServiceRegisterInverted);
+        if (pending == 0) {
+            return 8;
+        }
+
+        byte candidate = GetHighestPriorityIrq(pending);
+        if (candidate == 8) {
+            return 8;
+        }
+
+        if (IsSpecialMaskModeEnabled || HasHigherPriority(candidate, ActiveIrqLine)) {
+            return candidate;
+        }
+
+        return 8;
+    }
+
+    /// <summary>
+    ///     Updates the lowest-priority slot used to determine subsequent rotation order.
+    /// </summary>
+    /// <param name="value">IRQ index to treat as lowest priority.</param>
+    internal void SetLowestPriorityIrq(byte value) {
+        _lowestPriorityIrq = (byte)(value & 0x07);
+        if (Logger.IsEnabled(LogEventLevel.Debug)) {
+            Logger.Debug("PIC: Lowest priority set to IRQ {Irq}", _lowestPriorityIrq);
+        }
+    }
+
+    /// <summary>
+    ///     Updates the interrupt mask register and re-evaluates IRQs when newly masked bits intersect pending requests.
+    /// </summary>
+    /// <param name="value">Mask value written by the controller.</param>
+    public void SetInterruptMaskRegister(byte value) {
+        byte change = (byte)(InterruptMaskRegister ^ value); // Bits that have changed become 1.
+        InterruptMaskRegister = value;
+        InterruptMaskRegisterInverted = (byte)~value;
+        Logger.Debug("PIC: IMR updated to {Imr:X2} (changed {Changed:X2})", value, change);
+
+        // Test if changed bits are set in irr and are not being served at the moment.
+        // Those bits have an impact on whether the cpu emulation should be paused or not.
+        if ((InterruptRequestRegister & change & InServiceRegisterInverted) != 0) {
+            CheckForIrq();
+        }
+    }
+
+    /// <summary>
+    ///     Re-evaluates pending IRQs after an end-of-interrupt is issued.
+    /// </summary>
+    public void CheckAfterEoi() {
```

Comment:
You dont need that if you simplify the design meaning that IRQs are checked by the CPU when it needs them and not at other places.